File: Week02/Mod4-simpleskewness.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll  long long int
#define vi  vector<int>
#define vll vector<long long int>

double findmean(vll &pv, vll &v, int guess, int K, int N){
    if (K == 0)
        return v[guess];
    double sum;
    sum = (pv[guess+1] - pv[guess-K]) + (pv[N] - pv[N-K]); 
    return sum/(2*K+1);
}

int main(){
    #ifndef ONLINE_JUDGE 
        freopen("in.txt", "r", stdin);
        freopen("out.txt", "w", stdout);
    #endif

    std::ios::sync_with_stdio(false);

    int N;
    cin >> N;

    vll v; 
    int num;
    for(int i = 0; i < N; i++){
        cin >> num;
        v.push_back(num);
    }

    // Sort the array:
    
    sort(v.begin(), v.end());


    // Compute partial sums:

    vll pv;
    pv.push_back(0);
    for(int i = 0; i < N; i++){
        pv.push_back(pv.back()+v[i]);
    }


    // Sanity check, comment out later:

    // for(int i = 0; i < N; i++){
    //     cout << v[i] << "\t";
    // }

    // cout << "\n";

    // for(int i = 0; i < N; i++){
    //     cout << pv[i] << "\t";
    // }


    double answer = -1;
    double mean = 0, currmean = 0;
    int optguess, optK;

    for(int guess = 0; guess < N; guess++){

        // We want to pick K elements to the left of guess
        // and K elements from the end.
        // What's the largest possible value of K? 

        int range = min(guess, N-guess-1); 
        
        int L = 0, R = range, mid = (L+R)/2;

        while(R - L > 0){
            // mid is the current K.

            currmean = findmean(pv,v,guess,mid,N);
            // should we increase K? 

            int K = mid+1;
            int a = v[N-K];
            int b = v[guess-K];

            if(double(a+b)/2 <= currmean) // NO, K is an overestimate already.
                R = max(mid,L);
            else // YES, there's more potential for K! 
                L = min(mid + 1,R);     

            mid = (L+R)/2;
        }
        
        // When L = R, we have found K:

        mean = findmean(pv,v,guess,L,N);

        // If we are doing better than the best so far, update answers:

        if(mean - v[guess] > answer){
            answer = mean - v[guess];
            optguess = guess;
            optK = L;
        }
    }

    // Print answers:
    cout << optK*2 + 1 << "\n";
    
    for(int i = 0; i < optK; i++){
        cout << v[N-1-i] << " ";
    }
    
    for(int i = 0; i < optK+1; i++){
        cout << v[optguess-i] << " ";
    }
}

File: Week02/Mod4-simpleskewness.py
'''
Problem: Simple Skewness
URL: https://codeforces.com/contest/626/problem/E
Author : Ayush Modi
Discord Tag : hot9cups#3125
'''

from fractions import Fraction
# We need the fractions module because Python's inbuilt precision for floats isn't enough to solve the problem.
# Fraction is roughtly Python's equivalent of C++'s long double.

def findMean(prefixSum, n, k, medianIdx):
    total = prefixSum[medianIdx + 1] - prefixSum[medianIdx - k] + prefixSum[n] - prefixSum[n - k]
    return Fraction(total, (2 * k + 1))

n = int(input())
nums = sorted(list(map(int, input().split())))
bestSkewness, bestK, bestMedianIdx = 0, 0, 0

# Compute Prefix Sums:
prefSum = [0]
for i in range(n):
    prefSum.append(prefSum[-1] + nums[i])

for medianIdx in range(1, n-1):
    # We want to pick K elements to the left of medianIdx
    # and K elements from the end.
    # What's the largest possible value of K? 

    rangeOfK = min(medianIdx, n - medianIdx - 1)
    left, right = 0, rangeOfK
    
    # mid is the current K
    mid = (left + right) // 2
    
    while right > left:
        a = nums[n - mid - 1]
        b = nums[medianIdx - mid - 1]

        if (2 * mid + 1) * (a + b) <= 2 * (prefSum[medianIdx + 1] - prefSum[medianIdx - mid] + prefSum[n] - prefSum[n - mid]):
            right = mid     # K is an overestimate already.
        else:
            left = mid + 1  # There is more potential for K!
        mid = (left + right) // 2

    # When L = R, we have found K:
    bestCurrentMean = findMean(prefSum, n, left, medianIdx)
    # If we are doing better than the best so far, update answers:
    if bestCurrentMean - nums[medianIdx] > bestSkewness: 
        bestSkewness = bestCurrentMean - nums[medianIdx]
        bestMedianIdx = medianIdx
        bestK = left

# Print answers:
print(2 * bestK + 1)
for i in range(bestK):
    print(nums[n - 1 - i], end = " ")
for i in range(bestK+1):
    print(nums[bestMedianIdx - i], end = " ")
